ui/dashboards/src/context/TemplateVariableProvider.tsx

```diff
@@ -0,0 +1,267 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { createContext, useContext } from 'react';
+import { createStore, useStore } from 'zustand';
+import { immer } from 'zustand/middleware/immer';
+import { devtools } from 'zustand/middleware';
+import { TemplateVariableContext } from '@perses-dev/plugin-system';
+import {
+  VariableStateMap,
+  VariableState,
+  VariableName,
+  VariableValue,
+  ListVariableDefinition,
+  VariableDefinition,
+  DEFAULT_ALL_VALUE as ALL_VALUE,
+} from '@perses-dev/core';
+
+type TemplateVariableStore = {
+  variableDefinitions: VariableDefinition[];
+  variableState: VariableStateMap;
+  setVariableValue: (variableName: VariableName, value: VariableValue) => void;
+  loadTemplateVariable: (name: VariableName) => Promise<void>;
+};
+
+const TemplateVariableStoreContext = createContext<ReturnType<typeof createTemplateVariableSrvStore> | undefined>(
+  undefined
+);
+function useTemplateVariableStore() {
+  const context = useContext(TemplateVariableStoreContext);
+  if (!context) {
+    throw new Error('TemplateVariableStoreContext not initialized');
+  }
+  return context;
+}
+
+export function useTemplateVariableValues(variableNames?: string[]) {
+  const store = useTemplateVariableStore();
+  const state = useStore(
+    store,
+    (s) => {
+      const names = variableNames ?? Object.keys(s.variableState);
+      const vars: VariableStateMap = {};
+      names.forEach((name) => {
+        const varState = s.variableState[name];
+        if (!varState) {
+          return;
+        }
+        vars[name] = varState;
+      });
+      return vars;
+    },
+    (left, right) => {
+      return JSON.stringify(left) === JSON.stringify(right);
+    }
+  );
+  return state;
+}
+
+export function useTemplateVariable(name: string) {
+  const store = useTemplateVariableStore();
+  return useStore(store, (s) => {
+    const variableState = s.variableState[name];
+    const definition = s.variableDefinitions.find((v) => v.name === name);
+    return {
+      state: variableState,
+      definition,
+    };
+  });
+}
+
+export function useTemplateVariableActions() {
+  const store = useTemplateVariableStore();
+  return useStore(store, (s) => {
+    return {
+      setVariableValue: s.setVariableValue,
+      loadTemplateVariable: s.loadTemplateVariable,
+    };
+  });
+}
+
+export function useTemplateVariableDefinitions() {
+  const store = useTemplateVariableStore();
+  return useStore(store, (s) => s.variableDefinitions);
+}
+
+export function useTemplateVariableSrv() {
```

Comment:
Nit: Saw you renamed to remove the `Srv` part of the name (I'm having Angular flashbacks), might want to clean up a few more usages of `Srv` in here to reflect the new name?

Reply:
Good catch!